### ISSUE TITLE
Failed to link targets when building VASP+DFTD4

### ISSUE THREAD

**Issue by russell-samuel** [dftd4/dftd4#216]
**Describe the bug**
Failed to link targets when building VASP+DFTD4

**To Reproduce**

```bash
# 1. set environment
$ source /path/to/oneapi/setvars.sh
$ module load dft4_api

# 2. build vasp
$ cd /path/to/vasp
$ vim makefile.include
# 
# CPP_OPTIONS += -DDFTD4
# DFTD4_ROOT  ?= /path/to/dftd4/
# LLIBS       += -L$(DFTD4_ROOT)/lib64 -ldftd4
# INCS        += -I$(DFTD4_ROOT)/include -I$(DFTD4_ROOT)/include/dftd4/IntelLLVM-2023.1.0
# 
# LLIBS and INCS are the result of:
# pkg-config --libs dftd4_api  and  pkg-config --cflag dftd4_api 
# separately. 

$ make veryclean
$ make DEPS=1 -j
#
# the below is the make output
#
mpiifort -qmkl=sequential  # <-- here errors occurs when linking targets
   -o vasp c2f_interface.o ... subdftd3.o subdftd4.o ... main.o 
   -Llib -ldmy -Lparser -lparser -lstdc++ 
    -L/path/to/oneapi/mkl/2023.1.0/lib/intel64 
   -lmkl_scalapack_lp64 -lmkl_blacs_intelmpi_lp64 
   -L/path/to/hdf5/latest/lib -lhdf5_fortran 
   -L/path/to/dftd4/lib64 -ldftd4
# error output: 
/path/to/dftd4_api/lib64/libdftd4.a(compat.f90.o): In function `dftd4_compat_mp_d4_calculation_':
/path/to/dftd4_install/src/dftd4/compat.f90:111: undefined reference to `mctc_io_structure_mp_new_structure_num_'
/path/to/dftd4_install//3.6.0/src/dftd4/compat.f90:125: undefined reference to `mctc_io_math_mp_matinv_3x3_'
#
... # (Here I truncate the error output because too many 'undefined reference to')
#
/path/to/dftd4_api/lib64/libdftd4.a(charge.f90.o): In function `dftd4_charge_mp_get_charges_':
/path/to/dftd4_install/src/dftd4/charge.f90:57: undefined reference to `multicharge_param_mp_new_eeq2019_model_'
/path/to/dftd4_install//src/dftd4/charge.f90:58: undefined reference to `multicharge_cutoff_mp_get_lattice_points_cutoff_'
/path/to/dftd4_install//src/dftd4/charge.f90:65: undefined reference to /path/to/dftd4_install//src/dftd4/charge.f90:66: undefined reference to `multicharge_ncoord_mp_get_coordination_number_'
/path/to/dftd4_install//src/dftd4/charge.f90:68: undefined reference to `multicharge_model_mp_solve_'
/path/to/dftd4_install//lib64/libdftd4.a(charge.f90.o):(.data+0x10): undefined reference to `multicharge_model_mp_solve_'
make[2]: *** [makefile:132: vasp] Error 1
```

**Expected behaviour**

The linker can work properly when building VASP+DFTD4

**Additional context**

Using: 
- OneAPI 2023.1.0
- VASP 6.2.1
- DFTD4 3.6.0   (build from source with CMake 3.27.3)

I Googled and found some Q&A related to 'undefined reference to': 
- https://stackoverflow.com/questions/16088892/fortran-compilation-error-undefined-reference
- https://stackoverflow.com/questions/29639760/error-in-fortran-undefined-reference-to-subroutine

[According to IanH](https://stackoverflow.com/a/16093007) (a stackoverflow user), this may because: 
- Forget to `USE` module in main program `dft4.f90`
- the procedure in module is not `PUBLIC`

So I checked the source code of `dft4`, and nothing peculiar found. 

How I checked: 
Take the``charge.f90:57: undefined reference to `multicharge_param_mp_new_eeq2019_model_'`` in the last ``In function `dftd4_charge_mp_get_charges_': `` as an example: 
```bash
$ cd /path/to/dftd4_install

$ vim ./src/dftd4/charge.f90
# 
#  18 module dftd4_charge
# 
#  21    use multicharge, only : mchrg_model_type, new_eeq2019_model, &  
# 
#  27    public :: get_charges
# 
#  34 subroutine get_charges(mol, qvec, dqdr, dqdL)
#  35    !DEC$ ATTRIBUTES DLLEXPORT :: get_charges
# 
#  70 end subroutine get_charges
# 
#  73 end module dftd4_charge

$ vim ./subprojects/multicharge/src/multicharge/param.f90
#
# 16 module multicharge_param
#
# 25    public :: new_eeq2019_model
# 
# 29 subroutine new_eeq2019_model(mol, model)
# 
# 44 end subroutine new_eeq2019_model
#
# 46 end module multicharge_param
```
The `USE` and `PUBLIC` are properly used, hmmm, weird...

**Comment by awvwgk** [comment]
In this case it seems like CMake does not the dependencies on the libraries dftd4 needs for linking. You need to add `-lmulticharge -lmctc-lib` to the link line.

**Comment by russell-samuel** [comment]
Hmmm ... 🤔, I add these Libs and tried, still got error output.
Bad news is there is still an error occurs when linking. 
Good news is the error output is not the same. 

Here are the `makefile.include` and error outputs
```makefile
# makefile.include
CPP_OPTIONS += -DDFTD4
DFTD4_ROOT  ?= /opt/models/dftd4/3.6.0/
LLIBS       += -L$(DFTD4_ROOT)/lib64 -ldftd4 -lmulticharge -lmctc-lib -lmstore
INCS        += -I$(DFTD4_ROOT)/libdftd4.a.p
INCS        += -I$(DFTD4_ROOT)/libdftd4.a
INCS        += -I$(DFTD4_ROOT)/include \
               -I$(DFTD4_ROOT)/include/dftd4/IntelLLVM-2023.1.0 \
               -I$(DFTD4_ROOT)/include/mctc-lib/IntelLLVM-2023.1.0 \
               -I$(DFTD4_ROOT)/include/mstore/IntelLLVM-2023.1.0/ \
               -I$(DFTD4_ROOT)/include/multicharge/IntelLLVM-2023.1.0/ 
```

> NOTE: here in this setting existing a `libdftd4.a.p` as recommended on [VASP wiki](https://www.vasp.at/wiki/index.php/Makefile.include)
> but

**Comment by awvwgk** [comment]
You need to link with -fopenmp / -qopenmp

**Comment by russell-samuel** [comment]
It works 🎉🎉🎉

I could NOT successfully build it without your kindly help 👍👍👍

Thank you very much! 

=== === ===

**For another user who occurs this issue and don't know how to do**, here is part of `makefile.include`

```make
# makefile.include

CPP_OPTIONS += -D_OPENMP
FC           = mpiifort -qopenmp
FCL          = $(FC)

...

CPP_OPTIONS += -DDFTD4
DFTD4_ROOT  ?= /path/to/dftd4
LLIBS       += -L$(DFTD4_ROOT)/lib64 -ldftd4 -lmulticharge -lmctc-lib -lmstore
INCS        += -I$(DFTD4_ROOT)/libdftd4.a.p  # <-- NOTE the libdftd4.a.p is just the copy of libdftd4.a
INCS        += -I$(DFTD4_ROOT)/libdftd4.a
INCS        += -I$(DFTD4_ROOT)/include \
               -I$(DFTD4_ROOT)/include/dftd4/IntelLLVM-2023.1.0 \
               -I$(DFTD4_ROOT)/include/mctc-lib/IntelLLVM-2023.1.0 \
               -I$(DFTD4_ROOT)/include/mstore/IntelLLVM-2023.1.0/ \
               -I$(DFTD4_ROOT)/include/multicharge/IntelLLVM-2023.1.0/ 
```

> If one is enthusiastic about digging into the `-fopenmp`/`-qopenmp` option as @awvwgk mentioned, one can refer to: 
> - "MPI + OpenMP parallelization" section of [VASP wiki](https://www.vasp.at/wiki/index.php/Compiler_options)
> - [Intel Developer Guide Reference (English)](https://www.intel.com/content/www/us/en/docs/dpcpp-cpp-compiler/developer-guide-reference/2023-0/qopenmp-qopenmp.html)

**Comment by Andrew-S-Rosen** [comment]
@awvwgk: It might be worthwhile to add `-lmulticharge -lmctc-lib` to the installation instructions. I just ran into the same issue and was quite perplexed until I stumbled upon this (very helpful) thread. I was also using the OneAPI library from Intel.

My `makefile.include` looks like:

```
CPP_OPTIONS += -DDFTD4
LLIBS       += $(shell pkg-config --libs dftd4) -lmulticharge -lmctc-lib -lmstore
INCS        += $(shell pkg-config --cflags dftd4)
```

where `-lmulticharge -lmctc-lib -lmstore` are the additions compared to the installation instructions.
